crates/context_attribute/src/lib.rs

```diff
@@ -8,16 +8,36 @@ use syn::{
     spanned::Spanned,
     token::Mut,
     AngleBracketedGenericArguments, Expr, ExprLit, GenericArgument, GenericParam, ItemStruct,
-    Lifetime, LifetimeDef, Lit, Path, PathArguments, PathSegment, PredicateType, TraitBound,
-    TraitBoundModifier, Type, TypeParam, TypeParamBound, TypePath, TypeReference, WhereClause,
-    WherePredicate,
+    Lifetime, LifetimeDef, Lit, Path, PathArguments, PathSegment, Type, TypeParam, TypePath,
+    TypeReference,
 };
 
 #[proc_macro_attribute]
 #[proc_macro_error]
 pub fn context(_attributes: TokenStream, input: TokenStream) -> TokenStream {
     let mut struct_item = parse_macro_input!(input as ItemStruct);
 
+    let struct_name = struct_item.ident.to_string();
+    let allowed_member_types = match struct_name.as_str() {
+        "CreationContext" => ["Parameter", "PersistentState"].as_slice(),
+        "CycleContext" => [
+            "AdditionalOutput",
+            "HardwareInterface",
+            "HistoricInput",
+            "Input",
+            "Parameter",
+            "PerceptionInput",
+            "PersistentState",
+            "RequiredInput",
+        ]
+        .as_slice(),
+        "MainOutputs" => ["MainOutput"].as_slice(),
+        _ => abort!(
+            struct_item.ident,
+            "unexpected context name, try one of `CreationContext`, `CycleContext`, `MainOutputs`"
+        ),
+    };
```

Comment:
Deduplicate with the function in `source_analyzer`s `contexts.rs`?

Reply:
I tried that, but proc macro crates cannot export anything other than proc macros and having the context attribute crate depend on source analyzer didn't seem right to me either. Do you think that would be better?